src/gleam/list.gleam

```diff
@@ -454,6 +454,64 @@ pub fn fold_right(list: List(a), from initial: b, with fun: fn(a, b) -> b) -> b
   }
 }
 
+/// Like fold but the folding function also receives the index of the current element.
+///
+/// ## Examples
+///
+/// ```
+/// ["a", "b", "c"]
+/// |> list.index_fold([], fn(index, item, acc) { ... })
+/// ```
+///
+pub fn index_fold(
+  over over: List(a),
+  from initial: b,
+  with fun: fn(Int, a, b) -> b,
+) -> b {
+  over
+  |> index_map(fn(ix, value) { tuple(ix, value) })
```

Comment:
Could we do this in a single pass of the list rather than using `index_map`? If we're having a dedicated function for this I think we should aim to have it implemented more efficiently than using `index_map` followed by `fold`, as the user would do without this function.

Reply:
Done